### PR TITLE
[expo-updates][android] Add rollback to embedded update directive

### PR DESCRIPTION
**Pull request by wschurman**
# Why

Part 2 of https://github.com/expo/expo/pull/20275. This adds the new "rollback to embedded update" directive.

This adds support for a new directive type, `rollbackToEmbedded` - compliant servers may send this back to direct expo-updates to roll back to the embedded update.

# How

Roll back to embedded is accomplished by finding the embedded update in the database and setting it's commit time to that of the rollback message. This will have the effect of launching the embedded update iff there isn't a newer update in the db. Subsequent published updates will have later commit times and will take precedence over the embedded (until another rollback is published).

Some questions:
- Does this risk corruption of embedded updates? (permanently mutating it on the client)

# Test Plan

Run tests.

Manually test rollbacks by doing the following:
1. Implement the new protocol in https://github.com/expo/custom-expo-updates-server/pull/15
    - This links expo-updates in package.json to the local copy (this PR) `../../expo/packages/expo-updates` 
    - This also sets up `bundleInDebug: true` and `devDisabledInDebug: true` so that it loads updates in debug mode.
3. `npx expo prebuild` the example client.
4. Open the example client in android studio (verify it has new code from this PR)
5. Run the example server.
6. Build and run the client.
7. See embedded update on the client.
8. `yarn expo-publish`, kill and restart running client, see new update is downloaded and run
9. `yarn expo-publish-rollback`, kill and restart running client, see embedded update is run
10. `yarn expo-publish`, kill and restart running client, see new update is downloaded and run

# Checklist

- [ ] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [ ] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [ ] This diff will work correctly for `expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

**Comment by wschurman** [comment]
Copying relevant discussion from PR this was split out from:

> I still need to read over the code but I read the PR description. My initial thought is it would be more robust to install a new update from the embedded manifest, similar to when an app is newly downloaded and launched for the first time. There are two reasons for this. The most significant is that I recall we store only the current and previous updates in the database, and prune the rest. The entry for the embedded update might get cleared from the database and we won't be able to look up. The second reason is that out of principle I think it is better not to mutate updates in the database, since I believe we treat commitTimes as immutable everywhere else. That said, it's just out of principle and maybe it's tidier and more reliable to neatly update the commitTime.

In the discussion document we detailed a couple technical approaches to this: https://www.notion.so/expo/Expo-Updates-directives-in-multipart-manifests-2dee4deac86241a4873ee2fdb3d83d9c#a1426dba1f9348298567c16a4467a2ad

For installing a new update, we have the problem of duplicate manifest primary keys (IDs), so we'd need to delete the embedded update from the database if it exists and then replace it with a new copy. This presents a problem when the update being run is the embedded update, so we'd need to handle that case in a special way. 

To address the concern of the update having been pruned, this PR ensures the embedded update exists in the DB and then updates the timestamp (I believe, worth checking).

On the mutation note, we do a decent number of mutations to update objects in the updates db (status, asset status, keep) so that pattern exists. I do agree that mutating a piece of metadata originally derived from the embedded file is probably a bit fragile. I should look into how it is set for the embedded update in the first place (I think it is in a field of the embedded manifest and falls back to the current time, so changing it here could potentially create a mismatch between manifest state and db state).

**Review by douglowder — CHANGES_REQUESTED** [review]
See inline

File: packages/expo-updates/src/Updates.types.ts [hunk 1]
```diff
@@ -97,10 +97,23 @@ type UpdateFetchResultFailure = {
   manifest: undefined;
 };
 
+/**
+ * The rollback to embedded result of fetching a new update.
+ */
+type UpdateFetchResultRollbackToEmbedded = {
+  /**
+   * Signifies that the update was a roll back to the embedded update.
```
Comment (douglowder) [inline packages/expo-updates/src/Updates.types.ts:105]:
```suggestion
   * Signifies that the update was a rollback to the embedded update.
```

**Review by douglowder — APPROVED** [review]

File: packages/expo-updates/src/Updates.types.ts (hunk 1 above)
Comment (wschurman) [reply packages/expo-updates/src/Updates.types.ts:105]: We decided somewhere previously that "roll" and "back" should be different words (`RollBackToEmbeddedUpdateDirective`).

**Pull request merged**